File: src/slack/events.rs
use std::sync::Arc;
use std::time::Instant;
use tokio::sync::Mutex;
use tracing::{info, warn};

use super::client::SlackClient;
use super::format::md_to_slack;
use super::streamer::stream_to_slack;
use super::types::{SlackEvent, ThreadMessage};
use crate::claude::client::ClaudeClient;
use crate::claude::session::SessionStore;
use crate::claude::types::Message;
use crate::metrics::Metrics;
use crate::plugins::executor::execute_plugin;
use crate::plugins::registry::PluginRegistry;
use crate::plugins::router::route;

// ── Mention handler ─────────────────────────────────────────────────────────

/// Called for every `app_mention` event.
/// Runs the full plugin-routing + Claude-streaming flow.
pub async fn handle_mention(
    event: SlackEvent,
    slack: Arc<SlackClient>,
    claude: Arc<ClaudeClient>,
    sessions: Arc<Mutex<SessionStore>>,
    registry: Arc<PluginRegistry>,
    metrics: Arc<Metrics>,
    default_plugin: Option<String>,
) {
    let start = Instant::now();

    // 1. Extract channel.
    let channel = match &event.channel {
        Some(c) => c.clone(),
        None => return,
    };

    // 2. thread_ts: use existing thread, or fall back to the event's own ts.
    let thread_ts = event
        .thread_ts
        .clone()
        .or_else(|| event.ts.clone())
        .unwrap_or_default();

    let user = event.user.clone();

    // 3. Strip bot mention tokens and trim whitespace.
    let user_text = strip_mentions(event.text.as_deref().unwrap_or(""))
        .trim()
        .to_string();

    // Ignore empty messages.
    if user_text.is_empty() {
        return;
    }

    // Record mention in metrics.
    metrics.record_mention(user.as_deref(), &channel).await;

    // ── Built-in commands ────────────────────────────────────────────────

    // "stop" command
    if user_text.eq_ignore_ascii_case("stop") {
        let _ = slack
            .post_message(&channel, "🛑 Stopped.", Some(&thread_ts))
            .await;
        return;
    }

    // "stats" command
    if user_text.eq_ignore_ascii_case("stats") {
        let stats = metrics.format_stats().await;
        let _ = slack
            .post_message(&channel, &stats, Some(&thread_ts))
            .await;
        return;
    }

    info!(
        channel = %channel,
        thread_ts = %thread_ts,
        user = ?event.user,
        text = %user_text,
        "Handling app_mention"
    );

    // 4. Resolve external references in the message.
    let user_text = resolve_slack_urls(&user_text, &slack).await;
    let user_text = resolve_devrev_tickets(&user_text).await;

    // 5. Fetch thread history from Slack (source of truth) and build context.
    let thread_msgs = match slack.thread_history(&channel, &thread_ts).await {
        Ok(r) => r.messages.unwrap_or_default(),
        Err(e) => {
            warn!(error = %e, "Failed to fetch thread history");
            vec![]
        }
    };

    let thread_context = build_thread_context(&thread_msgs, event.ts.as_deref());

    // 6. Build session key and record user message (keeps session store in sync).
    let session_key = SessionStore::key(&channel, event.thread_ts.as_deref());
    {
        let mut store = sessions.lock().await;
        store
            .get_or_create(&session_key)
            .push(Message::user(&user_text));
    }

    // 7. Route to a plugin.
    let matched_plugin_fqn: Option<String> = if default_plugin.is_some() {
        default_plugin
    } else {
        route(&user_text, &registry).map(|p| p.fqn.clone())
    };

    if let Some(plugin_fqn) = matched_plugin_fqn {
        // ── Plugin path ─────────────────────────────────────────────────
        let plugin = match registry.get(&plugin_fqn) {
            Some(p) => p,
            None => {
                warn!(plugin = %plugin_fqn, "Plugin not found in registry");
                metrics.record_error();
                return;
            }
        };

        metrics.record_plugin_call(&plugin.fqn).await;

        // Post a "running plugin" indicator.
        if let Err(e) = slack
            .post_message(
                &channel,
                &format!("🔍 Running plugin: `{}`...", plugin.fqn),
                Some(&thread_ts),
            )
            .await
        {
            warn!(error = %e, "Failed to post plugin indicator");
        }

        // Enrich the query with thread context.
        let enriched_query = if thread_context.is_empty() {
            user_text.clone()
        } else {
            format!(
                "<thread_history>\n{}\n</thread_history>\n\n{}",
                thread_context, user_text
            )
        };

        // Execute the plugin with a single-turn call.
        let result =
            execute_plugin(plugin, &enriched_query, &[], &claude, &registry, 0).await;

        let reply_text = match &result {
            Ok(text) if text.is_empty() => "(no response from plugin)".to_string(),
            Ok(text) => {
                info!(len = text.len(), "Plugin returned response");
                text.clone()
            }
            Err(e) => {
                warn!(error = %e, "Plugin execution failed");
                metrics.record_error();
                format!("❌ Plugin error: {e}")
            }
        };

        let slack_text = md_to_slack(&reply_text);
        let post_resp = slack
            .post_message(&channel, &slack_text, Some(&thread_ts))
            .await;

        // Register bot message for reaction tracking.
        if let Ok(resp) = &post_resp {
            if let Some(ts) = &resp.ts {
                metrics
                    .register_bot_message(
                        &channel,
                        ts,
                        &thread_ts,
                        Some(&plugin.fqn),
                        user.as_deref(),
                        &user_text,
                    )
                    .await;
            }
        }
        if let Err(e) = post_resp {
            warn!(error = %e, "Failed to post plugin result");
        }

        // Record timing.
        metrics.record_response_time(start.elapsed()).await;

        // Record assistant response in session store.
        let mut store = sessions.lock().await;
        store
            .get_or_create(&session_key)
            .push(Message::assistant(&reply_text));
    } else {
        // ── Streaming path (no plugin) ──────────────────────────────────

        // Build system prompt with thread context.
        let system = if thread_context.is_empty() {
            None
        } else {
            Some(format!(
                "Below is the conversation history from this Slack thread. \
                 Use it as context to understand the ongoing discussion and \
                 provide a relevant response to the latest message.\n\n\
                 <thread_history>\n{}\n</thread_history>",
                thread_context
            ))
        };

        let messages = vec![Message::user(&user_text)];

        let rx = match claude.stream(messages, system.as_deref(), None).await {
            Ok(rx) => rx,
            Err(e) => {
                warn!(error = %e, "Claude stream failed to start");
                metrics.record_error();
                let _ = slack
                    .post_message(&channel, &format!("❌ Error: {e}"), Some(&thread_ts))
                    .await;
                return;
            }
        };

        let final_text = match stream_to_slack(
            rx,
            Arc::clone(&slack),
            channel.clone(),
            thread_ts.clone(),
            Arc::clone(&metrics),
            user.clone(),
            user_text.clone(),
        )
        .await
        {
            Ok(t) => t,
            Err(e) => {
                warn!(error = %e, "stream_to_slack failed");
                metrics.record_error();
                return;
            }
        };

        // Record timing.
        metrics.record_response_time(start.elapsed()).await;

        // Record assistant response in session store.
        if !final_text.is_empty() {
            let mut store = sessions.lock().await;
            store
                .get_or_create(&session_key)
                .push(Message::assistant(&final_text));
        }
    }
}

// ── Reaction handler ────────────────────────────────────────────────────────

/// Called for every `reaction_added` event.
/// Handles 👍/👎 feedback tracking.
pub async fn handle_reaction(event: SlackEvent, metrics: Arc<Metrics>) {
    let reaction = match &event.reaction {
        Some(r) => r.as_str(),
        None => return,
    };
    let item = match &event.item {
        Some(i) => i,
        None => return,
    };
    let item_channel = match &item.channel {
        Some(c) => c.as_str(),
        None => return,
    };
    let item_ts = match &item.ts {
        Some(t) => t.as_str(),
        None => return,
    };
    let user = event.user.as_deref().unwrap_or("unknown");

    metrics
        .record_reaction(item_channel, item_ts, reaction, user)
        .await;
}

// ── Helpers ─────────────────────────────────────────────────────────────────

/// Parse Slack thread URLs from a message and fetch their content.
async fn resolve_slack_urls(message: &str, slack: &SlackClient) -> String {
    let urls = extract_slack_urls(message);
    if urls.is_empty() {
        return message.to_string();
    }

    let mut context_parts: Vec<String> = Vec::new();

    for (channel_id, thread_ts) in &urls {
        info!(channel = %channel_id, ts = %thread_ts, "Fetching linked Slack thread");
        match slack.thread_history(channel_id, thread_ts).await {
            Ok(resp) => {
                if let Some(messages) = resp.messages {
                    let mut thread_text = String::new();
                    for msg in &messages {
                        let user = msg.user.as_deref().unwrap_or("unknown");
                        let text = msg.text.as_deref().unwrap_or("");
                        if !text.is_empty() {
                            thread_text.push_str(&format!("[{}]: {}\n", user, text));
                        }
                    }
                    if !thread_text.is_empty() {
                        context_parts.push(format!(
                            "--- Slack thread (#{}, ts: {}) ---\n{}---",
                            channel_id, thread_ts, thread_text
                        ));
                    }
                }
            }
            Err(e) => {
                warn!(channel = %channel_id, ts = %thread_ts, error = %e, "Failed to fetch linked thread");
            }
        }
    }

    if context_parts.is_empty() {
        return message.to_string();
    }

    format!(
        "{}\n\nUser's question: {}",
        context_parts.join("\n\n"),
        message
    )
}

fn extract_slack_urls(message: &str) -> Vec<(String, String)> {
    let mut results = Vec::new();
    let message = message.replace(['<', '>'], " ");

    for word in message.split_whitespace() {
        if !word.contains(".slack.com/archives/") {
            continue;
        }

        let parts: Vec<&str> = word.split('/').collect();
        if let Some(archives_idx) = parts.iter().position(|&p| p == "archives") {
            if archives_idx + 2 < parts.len() {
                let channel_id = parts[archives_idx + 1].to_string();
                let raw_ts = parts[archives_idx + 2];
                let raw_ts = raw_ts.split('?').next().unwrap_or(raw_ts);

                if let Some(ts_digits) = raw_ts.strip_prefix('p') {
                    if ts_digits.len() >= 7 {
                        let dot_pos = ts_digits.len() - 6;
                        let thread_ts =
                            format!("{}.{}", &ts_digits[..dot_pos], &ts_digits[dot_pos..]);
                        results.push((channel_id, thread_ts));
                    }
                }
            }
        }
    }

    results
}

async fn resolve_devrev_tickets(message: &str) -> String {
    let token = match std::env::var("DEVREV_TOKEN") {
        Ok(t) if !t.is_empty() => t,
        _ => return message.to_string(),
    };

    let ticket_ids = extract_devrev_ticket_ids(message);
    if ticket_ids.is_empty() {
        return message.to_string();
    }

    let http = reqwest::Client::new();
    let mut context_parts: Vec<String> = Vec::new();

    for ticket_id in &ticket_ids {
        info!(ticket = %ticket_id, "Fetching DevRev ticket");
        match fetch_devrev_ticket(&http, &token, ticket_id).await {
            Ok(content) => context_parts.push(content),
            Err(e) => {
                warn!(ticket = %ticket_id, error = %e, "Failed to fetch DevRev ticket");
            }
        }
    }

    if context_parts.is_empty() {
        return message.to_string();
    }

    format!(
        "{}\n\nUser's question: {}",
        context_parts.join("\n\n"),
        message
    )
}

fn extract_devrev_ticket_ids(message: &str) -> Vec<String> {
    let mut results = Vec::new();
    let message = message.replace(['<', '>'], " ");

    for word in message.split_whitespace() {
        let upper = word.to_uppercase();
        let cleaned = upper.trim_matches(|c: char| !c.is_alphanumeric() && c != '-');

        if (cleaned.starts_with("TKT-") || cleaned.starts_with("ISS-")) && cleaned.len() > 4 {
            results.push(cleaned.to_string());
            continue;
        }

        if word.contains("devrev.ai") {
            for prefix in &["/works/", "/issue/", "/ticket/"] {
                if let Some(pos) = word.find(prefix) {
                    let after = &word[pos + prefix.len()..];
                    let id = after
                        .split(|c: char| !c.is_alphanumeric() && c != '-')
                        .next()
                        .unwrap_or("")
                        .to_uppercase();
                    if (id.starts_with("TKT-") || id.starts_with("ISS-")) && id.len() > 4 {
                        results.push(id);
                    }
                    break;
                }
            }
        }
    }

    results.sort();
    results.dedup();
    results
}

async fn fetch_devrev_ticket(
    http: &reqwest::Client,
    token: &str,
    ticket_id: &str,
) -> Result<String, String> {
    let resp = http
        .get("https://api.devrev.ai/works.get")
        .header("Authorization", token)
        .query(&[("id", ticket_id)])
        .send()
        .await
        .map_err(|e| format!("DevRev request failed: {e}"))?;

    let status = resp.status();
    let body: serde_json::Value = resp
        .json()
        .await
        .map_err(|e| format!("DevRev response parse failed: {e}"))?;

    if !status.is_success() {
        return Err(format!(
            "DevRev returned {}: {}",
            status,
            body.get("message")
                .and_then(|v| v.as_str())
                .unwrap_or("unknown error")
        ));
    }

    let work = &body["work"];
    let title = work.get("title").and_then(|v| v.as_str()).unwrap_or("No title");
    let description = work
        .get("body")
        .and_then(|v| v.as_str())
        .unwrap_or("No description");
    let stage = work
        .pointer("/stage/name")
        .and_then(|v| v.as_str())
        .unwrap_or("unknown");
    let priority = work
        .get("priority")
        .and_then(|v| v.as_str())
        .unwrap_or("unknown");
    let created_by = work
        .pointer("/created_by/display_name")
        .and_then(|v| v.as_str())
        .unwrap_or("unknown");

    Ok(format!(
        "--- DevRev Ticket {} ---\nTitle: {}\nStage: {}\nPriority: {}\nCreated by: {}\nDescription:\n{}\n---",
        ticket_id, title, stage, priority, created_by, description
    ))
}

fn build_thread_context(messages: &[ThreadMessage], exclude_ts: Option<&str>) -> String {
    let mut parts: Vec<String> = Vec::new();

    for msg in messages {
        if let (Some(ts), Some(exc)) = (msg.ts.as_deref(), exclude_ts) {
            if ts == exc {
                continue;
            }
        }

        let text = msg.text.as_deref().unwrap_or("").trim();
        if text.is_empty() {
            continue;
        }

        if is_bot_indicator(text) {
            continue;
        }

        let sender = if msg.bot_id.is_some() {
            "assistant"
        } else {
            msg.user.as_deref().unwrap_or("unknown")
        };

        parts.push(format!("[{}]: {}", sender, text));
    }

    parts.join("\n")
}

fn is_bot_indicator(text: &str) -> bool {
    text.starts_with("⏳")
        || text.starts_with("🔍 Running plugin")
        || text == "🛑 Stopped."
        || text.starts_with("❌ Error:")
        || text.starts_with("❌ Plugin error:")
}

fn strip_mentions(text: &str) -> String {
    let mut result = String::with_capacity(text.len());
    let mut chars = text.chars().peekable();

    while let Some(c) = chars.next() {
        if c == '<' && chars.peek() == Some(&'@') {
            while let Some(inner) = chars.next() {
                if inner == '>' {
                    break;
                }
            }
            if chars.peek() == Some(&' ') {
                chars.next();
            }
        } else {
            result.push(c);
        }
    }

    result
}

File: src/oac_slack_bot/slack/events.py
"""Slack event handlers — mention and reaction processing."""

from __future__ import annotations

import asyncio
import os
import re
import time
from typing import TYPE_CHECKING

import httpx
import structlog

from oac_slack_bot.claude.session import SessionStore
from oac_slack_bot.claude.types import Message
from oac_slack_bot.metrics import Metrics
from oac_slack_bot.plugins.executor import execute_plugin
from oac_slack_bot.plugins.registry import PluginRegistry
from oac_slack_bot.plugins.router import route
from oac_slack_bot.slack.format import md_to_slack
from oac_slack_bot.slack.streamer import stream_to_slack
from oac_slack_bot.slack.types import SlackEvent, ThreadMessage

if TYPE_CHECKING:
    from oac_slack_bot.claude.client import ClaudeClient
    from oac_slack_bot.guardrails import GuardrailsPipeline
    from oac_slack_bot.slack.client import SlackClient

logger = structlog.get_logger()


async def handle_mention(
    event: SlackEvent,
    slack: SlackClient,
    claude: ClaudeClient,
    sessions: SessionStore,
    sessions_lock: asyncio.Lock,
    registry: PluginRegistry,
    metrics: Metrics,
    default_plugin: str | None,
    guardrails: GuardrailsPipeline | None = None,
) -> None:
    """Handle an app_mention event."""
    start = time.monotonic()

    channel = event.channel
    if not channel:
        return

    thread_ts = event.thread_ts or event.ts or ""
    user = event.user
    user_text = strip_mentions(event.text or "").strip()

    if not user_text:
        return

    await metrics.record_mention(user, channel)

    # Built-in commands
    if user_text.lower() == "stop":
        await slack.post_message(channel, "🛑 Stopped.", thread_ts)
        return

    if user_text.lower() == "stats":
        stats = await metrics.format_stats()
        await slack.post_message(channel, stats, thread_ts)
        return

    logger.info(
        "handling_app_mention",
        channel=channel,
        thread_ts=thread_ts,
        user=user,
        text=user_text[:200],
    )

    # Guardrails checkpoint
    if guardrails:
        result = await guardrails.check_input(user_text, user, channel)
        if not result.allowed:
            await slack.post_message(
                channel, f"⚠️ {result.reason}", thread_ts
            )
            return

    # Concurrent external resolution
    slack_resolved, devrev_resolved = await asyncio.gather(
        resolve_slack_urls(user_text, slack),
        resolve_devrev_tickets(user_text),
    )
    user_text = merge_resolved(user_text, slack_resolved, devrev_resolved)

    # Fetch thread history
    try:
        reply_resp = await slack.thread_history(channel, thread_ts)
        thread_msgs = reply_resp.messages or []
    except Exception as e:
        logger.warning("thread_history_failed", error=str(e))
        thread_msgs = []

    thread_context = build_thread_context(thread_msgs, exclude_ts=event.ts)

    # Session
    session_key = SessionStore.key(channel, event.thread_ts)
    async with sessions_lock:
        session = sessions.get_or_create(session_key)
        session.push(Message.user(user_text))

    # Route to plugin
    matched_fqn: str | None = default_plugin if default_plugin else None
    if not matched_fqn:
        matched = route(user_text, registry)
        matched_fqn = matched.fqn if matched else None

    if matched_fqn:
        # ── Plugin path ──
        plugin = registry.get(matched_fqn)
        if not plugin:
            logger.warning("plugin_not_found", plugin=matched_fqn)
            metrics.record_error()
            return

        await metrics.record_plugin_call(plugin.fqn)

        await slack.post_message(
            channel, f"🔍 Running plugin: `{plugin.fqn}`...", thread_ts
        )

        enriched_query = user_text
        if thread_context:
            enriched_query = (
                f"<thread_history>\n{thread_context}\n</thread_history>\n\n{user_text}"
            )

        try:
            reply_text = await execute_plugin(plugin, enriched_query)
            if not reply_text:
                reply_text = "(no response from plugin)"
            else:
                logger.info("plugin_response", len=len(reply_text))
        except Exception as e:
            logger.warning("plugin_execution_failed", error=str(e))
            metrics.record_error()
            reply_text = f"❌ Plugin error: {e}"

        # Filter output if guardrails enabled
        if guardrails:
            reply_text, warnings = guardrails.filter_output(reply_text)
            for w in warnings:
                logger.warning("output_filter_warning", warning=w)

        slack_text = md_to_slack(reply_text)
        post_resp = await slack.post_message(channel, slack_text, thread_ts)

        if post_resp.ts:
            await metrics.register_bot_message(
                channel, post_resp.ts, thread_ts,
                plugin.fqn, user, user_text,
            )

        elapsed = time.monotonic() - start
        await metrics.record_response_time(elapsed)

        async with sessions_lock:
            sessions.get_or_create(session_key).push(Message.assistant(reply_text))

    else:
        # ── Streaming path (no plugin) ──
        system = None
        if thread_context:
            system = (
                "Below is the conversation history from this Slack thread. "
                "Use it as context to understand the ongoing discussion and "
                "provide a relevant response to the latest message.\n\n"
                f"<thread_history>\n{thread_context}\n</thread_history>"
            )

        messages = [Message.user(user_text)]

        try:
            queue = await claude.stream(messages, system)
        except Exception as e:
            logger.warning("claude_stream_failed", error=str(e))
            metrics.record_error()
            await slack.post_message(channel, f"❌ Error: {e}", thread_ts)
            return

        try:
            final_text = await stream_to_slack(
                queue, slack, channel, thread_ts, metrics, user, user_text
            )
        except Exception as e:
            logger.warning("stream_to_slack_failed", error=str(e))
            metrics.record_error()
            return

        elapsed = time.monotonic() - start
        await metrics.record_response_time(elapsed)

        if final_text:
            async with sessions_lock:
                sessions.get_or_create(session_key).push(Message.assistant(final_text))


async def handle_reaction(event: SlackEvent, metrics: Metrics) -> None:
    """Handle a reaction_added event."""
    reaction = event.reaction
    item = event.item
    if not reaction or not item:
        return

    item_channel = item.channel
    item_ts = item.ts
    if not item_channel or not item_ts:
        return

    user = event.user or "unknown"
    await metrics.record_reaction(item_channel, item_ts, reaction, user)


# ── Helpers ──


def strip_mentions(text: str) -> str:
    """Remove <@BOT_ID> mention tokens from text."""
    result: list[str] = []
    chars = list(text)
    i = 0

    while i < len(chars):
        if chars[i] == "<" and i + 1 < len(chars) and chars[i + 1] == "@":
            # Consume until >
            while i < len(chars) and chars[i] != ">":
                i += 1
            i += 1  # skip >
            # Skip one trailing space
            if i < len(chars) and chars[i] == " ":
                i += 1
        else:
            result.append(chars[i])
            i += 1

    return "".join(result)


async def resolve_slack_urls(message: str, slack: SlackClient) -> str:
    """Fetch content from Slack thread URLs found in the message."""
    urls = extract_slack_urls(message)
    if not urls:
        return message

    async def fetch_thread(channel_id: str, thread_ts: str) -> str | None:
        logger.info("fetching_linked_thread", channel=channel_id, ts=thread_ts)
        try:
            resp = await slack.thread_history(channel_id, thread_ts)
            if not resp.messages:
                return None
            parts: list[str] = []
            for msg in resp.messages:
                user = msg.user or "unknown"
                text = msg.text or ""
                if text:
                    parts.append(f"[{user}]: {text}")
            if parts:
                return (
                    f"--- Slack thread (#{channel_id}, ts: {thread_ts}) ---\n"
                    + "\n".join(parts)
                    + "\n---"
                )
        except Exception as e:
            logger.warning("fetch_linked_thread_failed", channel=channel_id, error=str(e))
        return None

    results = await asyncio.gather(
        *(fetch_thread(ch, ts) for ch, ts in urls)
    )

    context_parts = [r for r in results if r]
    if not context_parts:
        return message

    return f"{chr(10).join(context_parts)}\n\nUser's question: {message}"


def extract_slack_urls(message: str) -> list[tuple[str, str]]:
    """Parse Slack thread URLs from a message."""
    results: list[tuple[str, str]] = []
    cleaned = message.replace("<", " ").replace(">", " ")

    for word in cleaned.split():
        if ".slack.com/archives/" not in word:
            continue

        parts = word.split("/")
        try:
            archives_idx = parts.index("archives")
        except ValueError:
            continue

        if archives_idx + 2 >= len(parts):
            continue

        channel_id = parts[archives_idx + 1]
        raw_ts = parts[archives_idx + 2].split("?")[0]

        if raw_ts.startswith("p") and len(raw_ts) >= 8:
            ts_digits = raw_ts[1:]
            dot_pos = len(ts_digits) - 6
            thread_ts = f"{ts_digits[:dot_pos]}.{ts_digits[dot_pos:]}"
            results.append((channel_id, thread_ts))

    return results


async def resolve_devrev_tickets(message: str) -> str:
    """Fetch DevRev ticket details for ticket IDs found in the message."""
    token = os.environ.get("DEVREV_TOKEN", "")
    if not token:
        return message

    ticket_ids = extract_devrev_ticket_ids(message)
    if not ticket_ids:
        return message

    http = httpx.AsyncClient(timeout=15)

    async def fetch_ticket(ticket_id: str) -> str | None:
        logger.info("fetching_devrev_ticket", ticket=ticket_id)
        try:
            resp = await http.get(
                "https://api.devrev.ai/works.get",
                headers={"Authorization": token},
                params={"id": ticket_id},
            )
            if resp.status_code != 200:
                return None
            body = resp.json()
            work = body.get("work", {})
            title = work.get("title", "No title")
            description = work.get("body", "No description")
            stage = (work.get("stage") or {}).get("name", "unknown")
            priority = work.get("priority", "unknown")
            created_by = (work.get("created_by") or {}).get("display_name", "unknown")
            return (
                f"--- DevRev Ticket {ticket_id} ---\n"
                f"Title: {title}\nStage: {stage}\nPriority: {priority}\n"
                f"Created by: {created_by}\nDescription:\n{description}\n---"
            )
        except Exception as e:
            logger.warning("devrev_fetch_failed", ticket=ticket_id, error=str(e))
            return None

    results = await asyncio.gather(*(fetch_ticket(tid) for tid in ticket_ids))
    await http.aclose()

    context_parts = [r for r in results if r]
    if not context_parts:
        return message

    return f"{chr(10).join(context_parts)}\n\nUser's question: {message}"


def extract_devrev_ticket_ids(message: str) -> list[str]:
    """Extract DevRev ticket IDs (TKT-xxx, ISS-xxx) from a message."""
    results: list[str] = []
    cleaned = message.replace("<", " ").replace(">", " ")

    for word in cleaned.split():
        upper = word.upper()
        stripped = upper.strip("".join(c for c in upper if not c.isalnum() and c != "-"))

        if (stripped.startswith("TKT-") or stripped.startswith("ISS-")) and len(stripped) > 4:
            results.append(stripped)
            continue

        if "devrev.ai" in word:
            for prefix in ("/works/", "/issue/", "/ticket/"):
                pos = word.find(prefix)
                if pos >= 0:
                    after = word[pos + len(prefix) :]
                    id_match = re.match(r"[A-Za-z0-9-]+", after)
                    if id_match:
                        tid = id_match.group().upper()
                        if (tid.startswith("TKT-") or tid.startswith("ISS-")) and len(tid) > 4:
                            results.append(tid)
                    break

    results = sorted(set(results))
    return results


def build_thread_context(
    messages: list[ThreadMessage], exclude_ts: str | None = None
) -> str:
    """Build context string from thread messages."""
    parts: list[str] = []

    for msg in messages:
        if exclude_ts and msg.ts == exclude_ts:
            continue

        text = (msg.text or "").strip()
        if not text:
            continue

        if _is_bot_indicator(text):
            continue

        sender = "assistant" if msg.bot_id else (msg.user or "unknown")
        parts.append(f"[{sender}]: {text}")

    return "\n".join(parts)


def merge_resolved(original: str, slack_resolved: str, devrev_resolved: str) -> str:
    """Merge results from concurrent URL and ticket resolution."""
    slack_changed = slack_resolved != original
    devrev_changed = devrev_resolved != original

    if not slack_changed and not devrev_changed:
        return original
    if slack_changed and not devrev_changed:
        return slack_resolved
    if not slack_changed and devrev_changed:
        return devrev_resolved

    # Both changed — combine contexts
    slack_context = slack_resolved.rsplit("\n\nUser's question: ", 1)[0]
    devrev_context = devrev_resolved.rsplit("\n\nUser's question: ", 1)[0]
    return f"{slack_context}\n\n{devrev_context}\n\nUser's question: {original}"


def _is_bot_indicator(text: str) -> bool:
    return (
        text.startswith("⏳")
        or text.startswith("🔍 Running plugin")
        or text == "🛑 Stopped."
        or text.startswith("❌ Error:")
        or text.startswith("❌ Plugin error:")
    )
